File: src/core/state.py
import time
from datetime import datetime
from typing import Dict, List

from src.core.record_intent import RecordIntent


class ContainerState:
    def __init__(
        self,
        container_id: str,
        container_name: str,
        container_created: datetime,
        record_intents: List[RecordIntent],
        status: str,
    ):
        self.container_id = container_id
        self.container_name = container_name
        self.container_created = container_created
        self.record_intents = record_intents
        self.status = status  # "running", "removed"
        self.last_updated = time.time()

    def is_stale(self, ttl: float = 30.0) -> bool:
        return (time.time() - self.last_updated) > ttl


class StateTracker:
    def __init__(self) -> None:
        self._containers: Dict[str, ContainerState] = {}

    def upsert(
        self,
        container_id: str,
        container_name: str,
        container_created: datetime,
        record_intents: List[RecordIntent],
        status: str,
    ) -> None:
        self._containers[container_id] = ContainerState(
            container_id, container_name, container_created, record_intents, status
        )

    def mark_removed(self, container_id: str) -> None:
        if container_id in self._containers:
            self._containers[container_id].status = "removed"
            self._containers[container_id].last_updated = time.time()

    def get_all_desired_record_intents(self) -> List[RecordIntent]:
        result = []
        for state in self._containers.values():
            if state.status == "running":
                result.extend(state.record_intents)
        return result

    def remove_stale(self, ttl: float = 60.0) -> None:
        # Unused at the moment.
        # Usage will require a recurring docker check to update lists of running containers
        # but this is redundant based on the way we process events
        expired = [cid for cid, state in self._containers.items() if state.is_stale(ttl)]
        for cid in expired:
            del self._containers[cid]

File: internal/core/state.go
package core

import (
	"sync"
	"time"

	"github.com/auto-dns/docker-coredns-sync/internal/intent"
)

// ContainerState holds state derived from container events.
type ContainerState struct {
	ContainerID   string
	ContainerName string
	Created       time.Time
	LastUpdated   time.Time
	RecordIntents []*intent.RecordIntent
	Status        string // "running", "removed"
}

// StateTracker stores container state safely.
type StateTracker struct {
	mu         sync.RWMutex
	containers map[string]*ContainerState
}

// NewStateTracker creates a new tracker.
func NewStateTracker() *StateTracker {
	return &StateTracker{
		containers: make(map[string]*ContainerState),
	}
}

// Upsert inserts or updates the state for a container.
func (s *StateTracker) Upsert(containerID, containerName string, created time.Time, intents []*intent.RecordIntent, status string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.containers[containerID] = &ContainerState{
		ContainerID:   containerID,
		ContainerName: containerName,
		Created:       created,
		RecordIntents: intents,
		Status:        status,
		LastUpdated:   time.Now(),
	}
}

// MarkRemoved marks a container state as removed.
func (s *StateTracker) MarkRemoved(containerID string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if state, exists := s.containers[containerID]; exists {
		state.Status = "removed"
		state.LastUpdated = time.Now()
		return true
	}
	return false
}

// GetAllDesiredRecordIntents returns all record intents from running containers.
func (s *StateTracker) GetAllDesiredRecordIntents() []*intent.RecordIntent {
	s.mu.RLock()
	defer s.mu.RUnlock()
	var intents []*intent.RecordIntent
	for _, cs := range s.containers {
		if cs.Status == "running" {
			intents = append(intents, cs.RecordIntents...)
		}
	}
	return intents
}
